torchao/prototype/awq/example.py

```diff
@@ -0,0 +1,256 @@
+    tokenizer = AutoTokenizer.from_pretrained(repo_id)
+    model = AutoModelForCausalLM.from_pretrained(repo_id, torch_dtype=precision).eval().to(device)
+    print(f"Time to load model: {time.time() - t0:.02f} seconds")
+    if quant.startswith("awq"):
+        quant_dtype = quant.split("-")[1]
+        quant_dtype = getattr(torch, quant_dtype, torch.bfloat16)
+        print(f"running {quant_dtype} calibration")
+        t0 = time.time()
+        # insert observers to find average magnitude and calculate scales
+        insert_awq_observer_(model,validation_size, sequence_length, quant_dtype=quant_dtype, group_size=group_size)
+        calibration_data = get_calib_dataset(tokenizer=tokenizer, n_samples=calibration_size, block_size=sequence_length)
+        for batch in calibration_data:
+            model(batch.to(device))
+            batch.to("cpu")
+        print(f"time for calibration: {time.time() - t0:.02f} seconds")
+        
+        is_observed_linear = lambda m, fqn: isinstance(m, AWQObservedLinear)
+        if "hqq" in quant:
+            print(f"running awq-hqq quantization")
+            from torchao.quantization.quant_primitives import (
+                MappingType,
+                ZeroPointDomain,
+                _DTYPE_TO_QVALUE_BOUNDS,
+            )
+            from torchao.dtypes import to_affine_quantized_intx, TensorCoreTiledLayoutType
+            # example of using a different quantization function
+            def hqqint4(weight):
+                mapping_type = MappingType.ASYMMETRIC
+                block_size = (1, group_size)
+                target_dtype = torch.int32
+                quant_min = 0
+                quant_max = 15
+                eps = 1e-6
+                preserve_zero = False
+                zero_point_dtype = torch.bfloat16
+                zero_point_domain = ZeroPointDomain.FLOAT
+        
+                return to_affine_quantized_intx(weight, mapping_type, block_size, target_dtype, quant_min, quant_max, eps, zero_point_dtype=zero_point_dtype, preserve_zero=preserve_zero, zero_point_domain=zero_point_domain, layout_type=TensorCoreTiledLayoutType(inner_k_tiles=8), use_hqq=True)
+            t0 = time.time()
+            quantize_(model, awq_uintx(quant_dtype=quant_dtype, group_size = group_size, weight_quant_fn=hqqint4), is_observed_linear)
+        else:
+            print(f"running {quant_dtype} quantization")
+            t0 = time.time()
+            # use awq_uintx() to apply awq quantization
+            quantize_(model, awq_uintx(quant_dtype=quant_dtype, group_size = group_size), is_observed_linear)
+        print(f"time for quantization: {time.time() - t0:.02f} seconds")
+        if model_save_path is not None:
+            print(f"Saving model to {model_save_path}")
+            torch.save(model, model_save_path)
+    elif quant=="int4":
+        print("running int4 quantization")
+        quantize_(model, int4_weight_only(group_size=group_size))
+    elif quant=="hqq":
```

Comment:
nit: might be better to be more specific here, like `int4wo-hqq-<group_size>` etc.